File: src/aizk/graph/profiles.py
import asyncio
from collections import defaultdict
from collections.abc import Sequence
from itertools import batched
from typing import cast

from loguru import logger
from patos import FrozenFlexModel, FrozenModel
from pydantic import UUID5, UUID7, TypeAdapter
from sqlalchemy.dialects.postgresql import insert
from sqlmodel import select

from ..config import settings
from ..exceptions import NotVisibleError
from ..serving.embed import Embedder
from ..serving.extract import LLM
from ..store import Entity, Fact, Profile, Watermark
from ..store.engine import Session
from ..store.identity import User
from ..types import Scopes
from .models import ProfileReport

_uuid5_adapter = TypeAdapter(UUID5)


class ProfileGrounding(FrozenModel):
    """One entity and the ordered current facts that ground its profile."""

    subject_id: UUID5
    name: str
    statements: tuple[str, ...]


class ProfileDraft(FrozenModel):
    """A summarized profile ready for one bulk upsert."""

    subject_id: UUID5
    summary: str
    vector: tuple[float, ...]


class ProfileBuilder(FrozenFlexModel):
    """Load, summarize, embed, and store profiles in bounded database phases."""

    scopes: Scopes
    llm: LLM
    embed: Embedder

    async def snapshot(
        self, session: Session, subject_ids: list[UUID5] | None = None
    ) -> list[ProfileGrounding]:
        """Load entity names and all related current fact statements in two queries."""
        roster = select(Entity.Content.id, Entity.Content.name).where(
            Entity.Content.id.in_(select(Entity.Claim.content_id))
        )
        if subject_ids is not None:
            roster = roster.where(Entity.Content.id.in_(subject_ids))
        entities = dict((await session.exec(roster.order_by(Entity.Content.id))).all())
        statements: dict[UUID5, list[str]] = defaultdict(list)
        if entities:
            rows = await session.exec(Fact.Live.touching(entities, settings.profile_facts_k))
            for entity_id, statement in rows:
                statements[entity_id].append(statement)
        return [
            ProfileGrounding(
                subject_id=subject_id,
                name=name,
                statements=tuple(statements[subject_id]),
            )
            for subject_id, name in entities.items()
            if statements[subject_id]
        ]

    async def summarize(self, groundings: list[ProfileGrounding]) -> list[ProfileDraft]:
        """Summarize every grounding and embed all resulting profiles in one batch."""
        reports: list[ProfileReport] = []
        for group in batched(groundings, settings.profile_build_concurrency, strict=False):
            reports.extend(
                await asyncio.gather(
                    *(
                        self.llm.generate(
                            settings.profile_system,
                            f"Entity: {grounding.name}\n\nFacts:\n"
                            + "\n".join(f"- {statement}" for statement in grounding.statements),
                            ProfileReport,
                        )
                        for grounding in group
                    )
                )
            )
        vectors = (
            await self.embed.embed([report.summary for report in reports], mode="document")
            if reports
            else []
        )
        return [
            ProfileDraft(
                subject_id=grounding.subject_id,
                summary=report.summary,
                vector=tuple(vector),
            )
            for grounding, report, vector in zip(groundings, reports, vectors, strict=True)
        ]

    async def store(self, session: Session, drafts: list[ProfileDraft]) -> dict[UUID5, UUID7]:
        """Upsert a complete profile batch and return IDs keyed by subject."""
        if not drafts:
            return {}
        base_statement = insert(Profile).values(
            [
                {
                    "created_by": settings.system_user_id,
                    "scopes": sorted(self.scopes),
                    "subject_id": draft.subject_id,
                    "summary": draft.summary,
                    "embedding": list(draft.vector),
                }
                for draft in drafts
            ]
        )
        statement = base_statement.on_conflict_do_update(
            index_elements=["scopes", "subject_id"],
            set_={
                "summary": base_statement.excluded.summary,
                "embedding": base_statement.excluded.embedding,
            },
        ).returning(Profile.subject_id, Profile.id)
        rows = cast("Sequence[tuple[UUID5, UUID7]]", (await session.exec(statement)).all())
        return dict(rows)


async def build_profile(
    subject_id: UUID5,
    llm: LLM,
    embed: Embedder,
    scopes: Scopes | None = None,
) -> UUID7:
    """Rebuild one entity profile through short read and write transactions."""
    key = frozenset(scopes or (settings.system_user_id,))
    builder = ProfileBuilder(scopes=key, llm=llm, embed=embed)
    async with User.system(key) as session:
        groundings = await builder.snapshot(session, [subject_id])
    if not groundings:
        raise NotVisibleError(f"entity {subject_id} is not visible or has no facts")
    drafts = await builder.summarize(groundings)
    async with User.system(key) as session:
        profile_ids = await builder.store(session, drafts)
    logger.info(
        "built profile for entity {} from {} facts", subject_id, len(groundings[0].statements)
    )
    return profile_ids[subject_id]


async def refresh_profiles(
    llm: LLM,
    embed: Embedder,
    scopes: Scopes | None = None,
) -> int:
    """Rebuild every visible profile with one snapshot and one bulk write."""
    key = frozenset(scopes or (settings.system_user_id,))
    builder = ProfileBuilder(scopes=key, llm=llm, embed=embed)
    async with User.system(key) as session:
        groundings = await builder.snapshot(session)
    drafts = await builder.summarize(groundings)
    async with User.system(key) as session:
        await builder.store(session, drafts)
    logger.info("refreshed {} profiles in scope {}", len(drafts), key)
    return len(drafts)


async def refresh_dirty_profiles(
    llm: LLM,
    embed: Embedder,
    scopes: Scopes | None = None,
) -> int:
    """Rebuild one bounded dirty-entity batch and consume only its observed increments."""
    key = frozenset(scopes or (settings.system_user_id,))
    builder = ProfileBuilder(scopes=key, llm=llm, embed=embed)
    async with User.system(key) as session:
        counters = await Watermark.pending_refs(
            session,
            key,
            Watermark.Kind.entity_dirty,
            settings.profile_batch_size,
        )
        groundings = await builder.snapshot(
            session,
            [_uuid5_adapter.validate_python(ref) for ref in counters],
        )
    drafts = await builder.summarize(groundings)
    async with User.system(key) as session:
        await builder.store(session, drafts)
        await Watermark.consume(
            session,
            key,
            Watermark.Kind.entity_dirty,
            counters,
        )
    logger.info("refreshed {} dirty profiles in scope {}", len(drafts), key)
    return len(drafts)

File: archive/ts/src/lib/server/graph/profiles.ts
// Entity profiles, the TS twin of graph/profiles.py: load the claimed entity roster and
// every related current fact in two RLS-filtered queries as the system identity, summarize
// each entity with the profile prompt, embed the batch, then upsert the whole generation
// in one app-role write keyed on (scopes, subject_id).
import { sql } from 'drizzle-orm';

import { embed, structured } from '../serving';
import { settings } from '../settings';
import { asSystem, scopeKey, uuidArray, type Reporter } from './system';

interface ProfileReport {
	summary: string;
}

const PROFILE_SCHEMA = {
	name: 'ProfileReport',
	schema: {
		type: 'object',
		properties: {
			summary: {
				type: 'string',
				description: 'one static-plus-dynamic paragraph grounded only in the facts'
			}
		},
		required: ['summary'],
		additionalProperties: false
	}
};

const report: Reporter<ProfileReport> = (system, user) =>
	structured<ProfileReport>(system, user, PROFILE_SCHEMA);

interface Grounding {
	subjectId: string;
	name: string;
	statements: string[];
}

// Load entity names and all related current fact statements in two queries, the
// ProfileBuilder.snapshot twin. Raises when a requested subject is not visible.
const snapshot = async (key: string[], subjectIds?: string[]): Promise<Grounding[]> =>
	asSystem(key, async (tx) => {
		const filter = subjectIds
			? sql` and ec.id = any(${uuidArray(subjectIds)}::uuid[])`
			: sql``;
		const roster = (await tx.execute(sql`
			select ec.id, ec.name from entity_content ec
			where ec.id in (select content_id from entity_claim)${filter}
			order by ec.id
		`)) as unknown as { id: string; name: string }[];
		const entities = new Map(roster.map((row) => [row.id, row.name]));
		if (subjectIds) {
			const missing = subjectIds.filter((id) => !entities.has(id)).sort();
			if (missing.length)
				throw new Error(`entities ${missing.join(', ')} are not visible to build profiles`);
		}
		const statements = new Map<string, string[]>();
		if (entities.size) {
			const ids = uuidArray([...entities.keys()]);
			const rows = (await tx.execute(sql`
				select subject_id, object_id, statement from live_fact
				where subject_id = any(${ids}::uuid[]) or object_id = any(${ids}::uuid[])
				order by lower(recorded), id
			`)) as unknown as { subject_id: string; object_id: string | null; statement: string }[];
			for (const row of rows) {
				statements.set(row.subject_id, [...(statements.get(row.subject_id) ?? []), row.statement]);
				if (row.object_id && entities.has(row.object_id) && row.object_id !== row.subject_id)
					statements.set(row.object_id, [...(statements.get(row.object_id) ?? []), row.statement]);
			}
		}
		return [...entities].map(([subjectId, name]) => ({
			subjectId,
			name,
			statements: statements.get(subjectId) ?? []
		}));
	});

interface Draft {
	subjectId: string;
	summary: string;
	vector: string;
}

// Summarize every grounding and embed all resulting profiles in one batch.
const summarize = async (
	groundings: Grounding[],
	summarizer: Reporter<ProfileReport>
): Promise<Draft[]> => {
	const reports: ProfileReport[] = [];
	for (const grounding of groundings)
		reports.push(
			await summarizer(
				settings.profileSystem,
				`Entity: ${grounding.name}\n\nFacts:\n` +
					grounding.statements.map((statement) => `- ${statement}`).join('\n')
			)
		);
	const vectors = await embed(reports.map((item) => item.summary), 'document');
	return groundings.map((grounding, index) => ({
		subjectId: grounding.subjectId,
		summary: reports[index].summary,
		vector: vectors[index]
	}));
};

// Upsert a complete profile batch and return profile ids keyed by subject.
const store = async (key: string[], drafts: Draft[]): Promise<Map<string, string>> => {
	if (!drafts.length) return new Map();
	const payload = JSON.stringify(
		drafts.map((draft) => ({
			subject_id: draft.subjectId,
			summary: draft.summary,
			vector: draft.vector
		}))
	);
	return asSystem(key, async (tx) => {
		const rows = (await tx.execute(sql`
			insert into profile (id, created_by, scopes, subject_id, summary, embedding)
			select uuidv7(), ${settings.systemUserId}, ${uuidArray(key)}::uuid[],
				d.subject_id, d.summary, (d.vector)::halfvec
			from jsonb_to_recordset(${payload}::jsonb) as d(subject_id uuid, summary text, vector text)
			on conflict (scopes, subject_id) do update
				set summary = excluded.summary, embedding = excluded.embedding
			returning subject_id, id
		`)) as unknown as { subject_id: string; id: string }[];
		return new Map(rows.map((row) => [row.subject_id, row.id]));
	});
};

// Rebuild one entity profile through short read and write transactions, the build_profile
// twin, returning the profile id.
export const buildProfile = async (
	subjectId: string,
	scopes?: string[],
	summarizer: Reporter<ProfileReport> = report
): Promise<string> => {
	const key = scopeKey(scopes);
	const groundings = await snapshot(key, [subjectId]);
	const drafts = await summarize(groundings, summarizer);
	const profileIds = await store(key, drafts);
	return profileIds.get(subjectId)!;
};

// Rebuild every visible profile with one snapshot and one bulk write, the refresh_profiles
// twin, returning the count.
export const refreshProfiles = async (
	scopes?: string[],
	summarizer: Reporter<ProfileReport> = report
): Promise<number> => {
	const key = scopeKey(scopes);
	const groundings = await snapshot(key);
	const drafts = await summarize(groundings, summarizer);
	await store(key, drafts);
	return drafts.length;
};
